### PR TITLE
Add emitjni target

### PR DESCRIPTION
**Pull request by gongsu832**
Add --EmitJNI target (tested working with mnist and resnet50)
    - MainUtils
      * first shot at refactoring compileModuleToSharedLibrary
      * add setExecPath call to allow resolving runtime directory from onnx-mlir
        executable path when ONNX_MLIR_RUNTIME_DIR is not set. This allows
        tests to run without having to install onnx-mlir or to explicitly set
        ONNX_MLIR_RUNTIME_DIR
    - RtMemRef
      * add getDataSize for C (equivalent of size() for C++).
      * fix setStrides bug (setting sizes, not strides)
    - TestConv
      * _main_graph-*.so were filling up /tmp. Change to use fixed shared library
        in build directory

### CODE REVIEW

File: src/MainUtils.hpp [hunk 1]
```diff
@@ -37,20 +37,30 @@
 #include "mlir/Target/LLVMIR.h"
 #include "mlir/Transforms/Passes.h"
 
+namespace onnx_mlir {
+extern std::string kExecPath;
```
Comment (tjingrant) [inline src/MainUtils.hpp:41]: Is this actually needed? Seems like only reference to this variable are inside MainUtil.cpp, and in MainUtil.cpp, ExternalUtil.hpp is actually included, which contains the definition.

File: src/MainUtils.cpp [hunk 2]
```diff
@@ -41,6 +40,29 @@ llvm::Optional<std::string> getEnvVar(std::string name) {
...
+//     installed system wide, which is typically /usr/local/bin, this will
+//     correctly resolve to /usr/local/lib), otherwise
+//   - use CMAKE_INSTALL_PREFIX/lib, which is typically /usr/local/lib
+string getRuntimeDir() {
+  if (const auto &dir = getEnvVar("ONNX_MLIR_RUNTIME_DIR"))
+    return dir.getValue();
+
+  string dir = llvm::sys::path::parent_path(kExecPath).str();
+  if (llvm::sys::path::stem(dir).str().compare("bin") == 0)
+    return dir.substr(0, dir.size() - 3) + "lib";
```
[the remark this answers is not in the document]
Comment (gongsu832) [reply src/MainUtils.cpp:59]: lib64 is for systems that still haven't manage to completely get rid of the 32-bit baggage. Eventually it's expected that everybody will move to lib. But agreed that we should still check for lib64. I will add that.

File: src/MainUtils.cpp [hunk 3]
```diff
@@ -204,42 +229,119 @@ void compileModuleToSharedLibrary(
...
+
+// Link everything into a shared object.
+void genSharedLib(const mlir::OwningModuleRef &module,
+    string modelSharedLibPath, std::vector<string> opts,
+    std::vector<string> objs, std::vector<string> libs) {
 
-  llvm::Optional<std::string> runtimeDirInclFlag;
-  if (getEnvVar("RUNTIME_DIR").hasValue())
-    runtimeDirInclFlag = "-L" + getEnvVar("RUNTIME_DIR").getValue();
+  llvm::Optional<std::string> runtimeDirInclFlag = "-L" + getRuntimeDir();
```
[the remark this answers is not in the document]
Comment (gongsu832) [reply src/MainUtils.cpp:268]: Yes we can remove `llvm::Optional`.

File: src/MainUtils.hpp (hunk 1 above)
Comment (gongsu832) [reply src/MainUtils.hpp:41]: I was hoping that we don't have to do this but unfortunately not everybody includes `ExternalUtil.hpp`. For example, `TestConv.cpp` includes `MainUtil.hpp` but not `ExternalUtil.hpp`. If there is some other better way to "fix" this, feel free to suggest.

Comment (gongsu832) [reply src/MainUtils.hpp]: Hmm...I think we can indeed remove it from `MainUtils.hpp`. At some point I was setting `kExecPath` directly in `main()` so `TestConv.cpp` won't compile without the `extern std::string kExecPath`. Then I made a `setExecPath()` call in `MainUtils.cpp` to do that and forgot that now we can indeed remove the `extern`.

Thanks for catching it!

Comment (tjingrant) [reply src/MainUtils.hpp]: Sounds good!

**Pull request merged**